### ISSUE TITLE
Question about tokenization_helpers.py

### ISSUE THREAD

**Issue by TahaBinhuraib** [HuthLab/encoding-model-scaling-laws#7]
Hi,

Thank you very much for making this repo public with the code!

I just have a question about the 'tokenization_helpers.py' file, I'm not able to follow the logic behind it fully. Would it be possible to generate a function that is tokenizer agnostic so that we could test different models? I think this would make much more sense even if it results in a slight drop in performance. I tried taking the approach by Schrimpf et al. But the performance drop was quite significant. I'm not sure if this is due to the boundaries between sentences which results in a loss of context. In any case, what would be the ideal way to test different models?

Thanks!

**Comment by RAntonello** [comment]
Hi,

Unfortunately it is somewhat challenging to provide an efficient and completely model-agnostic tokenizer owing to the the wide diversity in tokenization strategies. The challenge mostly arises from the differences in how the different strategies handle whitespace. However, allow me to instead provide you with a high-level description of the idea behind the tokenization. The problem that needs to be solved is that our temporal annotations exist at the word level, but the tokens themselves can occur along arbitrary character boundaries. We therefore need to reconcile these two (the annotations and the tokenization scheme) by additionally figuring out which tokens occur on the boundaries between words. The simplest way to do this algorithmically is to just apply a .split() to the string containing the story data, and then to run one forward pass through the model per word. So for instance, if the full story were just the phrase "i picked up eggs from the store", we would run the model once for each of the following strings

'i'
'i picked'
'i picked up'
'i picked up eggs'
'i picked up eggs from'
'i picked up eggs from the'
'i picked up eggs from the store'

Once we did that, we would take the last token embedding of the model and apply our lanczos interpolation and that would be all. If you have a surplus of GPU compute, I would recommend this approach. It can be formalized as something like this, which extracts all the hidden states for every layer:

```
def make_semantic_model_lm_layers_all(ds, model, tokenizer, num_layers, hidden_state_size):
    hidden_state_lst = []
    lookback = 64 # how long you want your max string size to be
    for i, w in enumerate(ds.data):
        if w != '':
            text = [" ".join(ds.data[max(0,i-lookback):min(i+1,total_len)])]
            tokens = tokenizer(text)['input_ids']
            out = torch.cat(list(model(torch.tensor(tokens))[1]), axis=0).cpu().detach().numpy()
            hidden_state_lst.append(out[:,-1,:])
        else:
            hidden_state_lst.append(np.zeros((num_layers,hidden_state_size)))
       return hidden_state_lst
```
       
This works although it is very inefficient, as it requires one forward pass per word. What we do differently in the tokenzation_helpers functions is basically some simple dynamic programming, we first go through the entire story, annotating which tokens correspond to word boundaries using a dummy token as a marker (e.g. 27 or 29947 in the opt and llama tokenization helpers), and then we figure out how to orient the forward passes so we can compute the hidden states for many words all at once in a single forward pass, if we allow a slight relaxation of the requirement that every context have the exact same length. This enables a large constant time speedup in the hidden state generation because you aren't constantly recomputing hidden states as in the above example, but it is more tedious to write because it involves keeping track of a lot of indices. I hope this is helpful to you.

**Comment by TahaBinhuraib** [comment]
Thank you so much!
